### PR TITLE
Fix FP in `single_component_path_imports` lint

### LINKED ISSUE #5210: FP single_component_path_imports
`clippy 0.0.212 (2855b21 2020-02-19)`

Cargo.toml:
````toml
...
[dependencies]
crypto-hash = "0.3.1"
````

main.rs:
````rust
use self::crypto_hash::{Algorithm, Hasher};
use crypto_hash;

fn main() {}

````

This will cause a warning:

````
warning: this import is redundant
 --> src/main.rs:2:1
  |
2 | use crypto_hash;
  | ^^^^^^^^^^^^^^^^ help: remove it entirely
  |
  = note: `#[warn(clippy::single_component_path_imports)]` on by default
````
but removing it will lead to a compiler error:
````
error[E0432]: unresolved import `self::crypto_hash`
 --> src/main.rs:1:11
  |
1 | use self::crypto_hash::{Algorithm, Hasher};
  |           ^^^^^^^^^^^ maybe a missing crate `crypto_hash`?

error: aborting due to previous error
````

### PR DESCRIPTION
**Pull request by ThibsG**
Fix FP in  `single_component_path_imports` lint when the import is reused with `self`, like in `use self::module`.

Fixes #5210 

changelog: none

### CODE REVIEW

File: tests/ui/single_component_path_imports_self.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// edition:2018
+#![warn(clippy::single_component_path_imports)]
+#![allow(unused_imports)]
+
+use self::regex::{Regex as xeger, RegexSet as tesxeger};
+pub use self::{
+    regex::{Regex, RegexSet},
+    some_mod::SomeType,
+};
+use regex;
```
Comment (giraffate) [inline tests/ui/single_component_path_imports_self.rs:10]: Does the test case below work?
```
use regex;

use self::regex::{Regex as xeger, RegexSet as tesxeger};
pub use self::{
    regex::{Regex, RegexSet},
    some_mod::SomeType,
};
```

Comment (ThibsG) [reply tests/ui/single_component_path_imports_self.rs:10]: Doesn't work, I guess I misinterpreted the meaning of `check_item_post` 😅

File: clippy_lints/src/single_component_path_imports.rs [hunk 2]
```diff
@@ -34,29 +36,87 @@ declare_clippy_lint! {
...
-        if_chain! {
-            if !in_macro(item.span);
-            if cx.sess.opts.edition >= Edition::Edition2018;
-            if !item.vis.kind.is_pub();
-            if let ItemKind::Use(use_tree) = &item.kind;
-            if let segments = &use_tree.prefix.segments;
-            if segments.len() == 1;
-            if let UseTreeKind::Simple(None, _, _) = use_tree.kind;
-            then {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, krate: &Crate) {
```
[the remark this answers is not in the document]
Comment (ThibsG) [reply clippy_lints/src/single_component_path_imports.rs:60]: Good catch, thanks !

File: tests/ui/single_component_path_imports.rs [hunk 3]
```diff
@@ -19,3 +19,9 @@ fn main() {
     // False positive #5154, shouldn't trigger lint.
     m!();
 }
+
+mod hello_mod {
+    use regex;
+    #[allow(dead_code)]
+    fn hello_mod() {}
+}
```
Comment (giraffate) [inline tests/ui/single_component_path_imports.rs:27]: Thanks for moving this forward!

I think we need to check it for each module. If the codes like below added to this test code, it looks like that the lint isn't emitted at all in this test.
```rust
mod hi_mod {
    use self::regex::{Regex, RegexSet};
    use regex;
    #[allow(dead_code)]
    fn hi_mod() {}
}
```

**Review by giraffate — APPROVED** [review]

**Comment by giraffate** [comment]
@bors r+

It looks good, thanks!

**Comment by giraffate** [comment]
Oh, the test fails because of ICE, I missed it. I will check it later.

File: clippy_lints/src/single_component_path_imports.rs [hunk 4]
```diff
@@ -38,26 +37,93 @@ declare_clippy_lint! {
...
+            if segments.len() == 1 {
+                if let UseTreeKind::Simple(None, _, _) = use_tree.kind {
+                    let ident = &segments[0].ident;
+                    single_use_usages.push((ident.name, item.span));
+                }
+                return;
+            }
+
+            // keep track of `use self::some_module` usages
+            if segments[0].ident.name == kw::SelfLower {
```
Comment (giraffate) [inline clippy_lints/src/single_component_path_imports.rs:109]: @ThibsG Can you address the pattern like below? ICE happen here.
```
mod root_nested_use_mod {
    use {regex, serde};
    #[allow(dead_code)]
    fn root_nested_use_mod() {}
}
```

**Review by giraffate — CHANGES_REQUESTED** [review]

File: clippy_lints/src/single_component_path_imports.rs (hunk 4 above)
Comment (ThibsG) [reply clippy_lints/src/single_component_path_imports.rs:109]: Ah interesting, I didn't even know the syntax `use {regex, serde};` is legit code.

I will add this test case.

**Comment by djc** [comment]
Did this also the fix the case of `use crate::foo`? That is, I have the following:

`lib.rs`:

```
#[cfg(foo)]
use log;

#[cfg(not(foo))]
#[macro_use]
mod log {
    macro_rules! error    ( ($($tt:tt)*) => {{}} );
};

mod inner {
    use crate::log::error;
}
```

**Comment by ThibsG** [comment]
@djc, no I don't think this case is handled, but feel free to provide a playground, and I will take a look at this.

**Review by giraffate — APPROVED** [review]

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**